### ISSUE TITLE
Autogenerated factories for parameterized constructors? 

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#49]
This is to discuss in details the offshoot from the discussion in #46. 

tB allows for parameterized construction of objects. That's great!

Unfortunately, COM cannot use those constructors directly.

However, we can still help COM by making that object `PublicNotCreatable` and then creating a factory (class? module?) that can then construct the object and return it, thus effectively encapsulating the construction neatly for COM consumers. 

It should be possible to then autogenerate a factory method for every parameterized constructor.

I'm inclined to create a factory module for each object that wants to be COM-visible but has parameterized constructors because:

1) construction shouldn't require yet another object. Module function that returns an object avoids the requirement to create factory class which results in more verbose and less readable code, IMO. 
2) construction generally shouldn't be stateful. This is not to say there won't be stateful factories. But for the purpose of autogenerated factories, it'll be just a thin wrapper of dispatching to the parameterized constructor so there should be no state. Module convey that idea more clearly than a class.
3) Having a module for each object (e.g. `MyClass` + `MyClassFactory`) seems more logical even though it may result in more clutters in the type library mainly because `GlobalFactory` may end up being too cluttered if it can create all unrelated objects.

**Comment by retailcoder** [comment]
In VBA I personally like having a factory method in the class it's creating an instance of, for several reasons:

1. Calling `Thing.Create` off the default instance looks & feels *exactly right*.
2. It puts construction and dependency injection concerns where they belong, in the class module being instantiated.
3. It separates the class' default/stateless interface from its explicit/stateful interface; instance members are explicitly defined in an explicit interface.

In twinBASIC, with parameterized constructors *factory methods* are much less commonly needed, but *abstract factories* remain a valid and useful tool for dependency injection. The way I see it, the need for tB factory methods is rather limited, as much as it is in C# or VB.NET, because a factory method necessarily induces often-undesirable *coupling*.

Thus, if an *Extract Factory* refactoring tool is to be made, it should offer several options:
- *Extract Factory Method*: in a class with `VB_PredeclaredId` and `VB_Exposed` set to `True` and `VB_Creatable` set to `False`, adds a `Public Function Create` that copies the signature of the selected/current constructor overload, invokes it, and returns an explicit interface (prompt/select, I guess) if the class has `Implements` statements (extracts an interface otherwise?), or returns the class' default interface otherwise. Now that's... a lot of conditions and constraints I think.
- *Extract Factory*: creates a new `VB_Creatable` class with only the extracted `Create` method/function, parameterized as per a selected constructor overload. Returns the class' default interface.
- *Extract Abstract Factory*: in a class with `Implements` statements, creates a new `VB_Creatable = False` interface class with a `Create` method/function parameterized as per the selected constructor overload, and returns one of the interfaces implemented by the class. Optionally, the extracted factory interface could be automatically implemented by invoking the selected/wrapped constructor overload.

Factories are OOP stuff. Any "extract factory" tool that doesn't involve a class module should explicitly mention it:

- *Extract to Factory Module*: creates a new `Create{ClassName}` function in a standard module copying the signature of the selected constructor overload, returns the class' default interface.

I personally profoundly dislike the idea of having a God-like `AllTheFactories` module or class that can spawn an instance of just about anything from just about anywhere. While it should be *possible* to do it in tB, I don't think tB should be providing tooling to *facilitate* it.

**Comment by bclothier** [comment]
I agree that having `Thing.Create(...)` is the best solution for most classes. My main problem is that there's no guard against accidentally using `Thing` as an instance when it should be used like a type. I am not sure that the concept of `Shared` (or C#'s `static`) works in COM where everything are interfaces or modules. Given the tB's mandate to be COM compatible, the solution has to be something that will work even for external COM consumers. 

I also am not sure it's possible for tB to guard against misuse of the default instance for predeclared instances. The best thing we can hope is for a runtime error when a default instance is misused. That's not very optimal. For that reason, the ideal solution would come up with something that will be possible for the COM consumers to correctly create the object without going through a default instance.

**Comment by WaynePhillipsEA** [comment]
If we're ruling out the god-like AllTheFactories class/module, then we're left with a per class solution.  Since we cannot expose real static (Shared) members on a class via COM, that means we need an extra factory-class (or module) for each real class that exposes parameterized constructors.  Along the lines of MyClass1_Factory.Create(...)?

**Comment by WaynePhillipsEA** [comment]
Or thinking outside of the box, we could generate a second type library (still embedded inside the same generated DLL), with a different typelibrary name.... just for the factories, something like:

    MyProjectFactories.MyClass1.Create()

**Comment by bclothier** [comment]
I do not know if this will help further the discussion but I had a situation where I had a complex construction requirement and set this up, borrowing the idea from Microsoft's `ICreateTypeLib` & `ITypeLib`:

```
Public Interface ICreateFoo
  Create()
End Interface

Public Interface IFoo
  DoFoo()
End Interface

Public Class Foo
  Implements ICreateFoo
  Implements IFoo

  Private Sub ICreateFoo_Create() As IFoo
  End Sub

  Private Sub IFoo_DoFoo()
  End Sub
End Class

Public Module FooFactory
  Public Sub Create() As IFoo
    Dim Foo As ICreateFoo
    Set Foo = New Foo
    Set Create = Foo.Create()
  End Sub
End Module
```

This is a _lot_ of boilerplate all just to work around the need to set up the private state of the instance via the `ICreateFoo` implemented on the class, then hide the private state by showing only the `IFoo` interface. Note that one can still defeat the encapsulation simply by casting `IFoo` back to `ICreateFoo`. Were I able to write the `IUnknown::QueryInterface`, I would probably looked at whether I could make the casting a one way (e.g. you can cast from `ICreateFoo` to `IFoo` but not from `IFoo` to `ICreateFoo`. However, we probably want the `IFoo` interface to be the default interface, yet it's going to be `ICreateFoo` that should be used first during the construction of the object. 

Again, all that boilerplate code was made necessary because of VBA's limitations. I am not proposing that we keep doing this in tB. I'd prefer a solution that leads to less boilerplate (more readability!) But if the design above is doable from COM's standpoint, maybe that may be helpful?

**Comment by Kr00l** [comment]
Could be a dumb question...

Is the VTable offset affected by this for a class?

For full VB6/VBA compatibility the offset should be "1Ch".
So the ASM thunks would still work to get the first routine after IDispatch of a class.

"call    dword [eax+1Ch]"

**Comment by WaynePhillipsEA** [comment]
@Kr00l no, nothing here should affect the vtables.  Plus I've got good test coverage on the vtables to ensure they always match the VB6/VBA produced versions.

**Comment by vbRichClient** [comment]
IMO we have two use-cases:
- twin-internal Constructors (which could remain "as they are and currently work")
- twin-compiled COM-Dlls (which - IMO - should support a factory-class)

The following will focus only onto that latter case...

What nobody mentioned so far, is:
- "regfree usage" of these twin-compiled COM-Libs
- which immediately brings to mind "platform-independent usage" of these "libs with Classes" (regfree-ness being a necessity)

To solve both of the above (in addition to constructor-support), I would:
- export an additional "flat-API-function", named GetFactory() 
- no Params to pass, but returning a concrete Factory-ClassType which is "Dll-specific" (knows all PublicCreatable (Co)Classes)

Users of a given Dll (no matter whether on Windows or Linux) could retrieve such a factory directly - 
(without cluttering or requiring "a registry"), assuming the COM-lib was named MyCOMLib.dll (or MyCOMLib.so) via:
- in VB6/VBA: Private Declare Function GetFactory Lib "MyCOMLib" () As MyCOMLib.cFactory
- on Non-Win-systems: via matching DyLib-load mechanisms and an accompanying MyCOMLib.h, or MyCOMLib.bi file)

So the cFactory-Class will contain "as many Constructor-Functions, as publically-creatable Classes exist in that Dll-Project".
In addition (to keep the amount of "Flat-APIs-to-Declare" low), cFactory could also implement an Interface IReflection:
- GetClasses ... (an enumerator-host-class, able to enumerate IClassDef-Members)
- GetStructs ... (dito here, for publically exposed UserDefined Types)
- GetEnums ... (and Public Enum-Defs)
- GetConsts ... (you get the idea)

This way (using such a "Flat-API-exposed Factory at Lib-level" - usable via "a simple DyLibLoad-Declare"), 
other languages (and other platforms) will be able, to not only:
- instantiate the Public-creatable flagged Dll-CoClasses with Constructor-functionality
- but also to use Reflection on all the publically visible Types of that COM-compatible Class-Library

For Windows-platforms, this concept will (also) allow "regfree usage of twin-produced COM-libs out of the box"...
And of course, these Libs should support the "classic" 4 Standard-Flat-API-exports of COM-libs:
- DllRegisterServer
- DllGetClassObject
- DllCanUnloadNow
- DllUnregisterServer
as well, to allow for "classic, registry-based instancing" on Windows-platforms in addition.

Just my $0.02

Olaf

**Comment by bclothier** [comment]
I love the idea of having a exported function to get factories. However I need to point out that in the case of VBA consumers, they must have a type library; they cannot use COM objects without a type library. They could late-bind using `Object` / `IDispatch` but that is not very fun to develop with. If I want my tB DLL to be adopted by VBA users, it needs to somehow register the type library. I've been doing this by using HKCU to avoid the issues with registry permissions but that is not really "reg-free". 

The ability to use tB DLL early-bound and not have to set a reference would be also valuable to the VBA users who may not want the hassle of registering it, even if we make it very easy. 

It's not clear to me how other languages, especially those on other platforms would use the objects if they don't have the COM support. It seems to me it implies a method to generate the equivalent of C++ header files so that the other language can know how to use it. In .NET, we would just create an `[ComImport(...)] public interface { ... }` to import a unknown COM type but that requires knowledge about the object's shape at the compile-time.

**Comment by vbRichClient** [comment]
> I need to point out that in the case of VBA consumers, they must have a type library...
Sure - also VB6-Users (or twin-Users who use a twin-Class-Dll on Windows) will need the typelib 
(either in the binary itself - or in a separate *.tlb file). 

Instancing a factory-class regfree (via an exported Flat-API like GetFactory()) is one thing.

Knowing the vtable-members (and the method-signatures) of the returned cFactory - 
(as well as all the other Public Classes within that COM-lib) beforehand, is another.

Well, for latebound usage a consuming client only needs to know, how to deal with IDispatch - 
but for early-binding on Windows, typelibs are paramount.

> It's not clear to me how other languages, especially those on other platforms would use the objects if they don't have COM support.
For C-style consumers (e.g. on Linux), a normal C-HeaderFile should be sufficient, 
e.g. with Struct-Definitions which contains the C-style function-declarations (to reflect the VTable for a given Class).

And I like the idea, of another Flat-API-export for Linux *.so libs, as e.g. GetHeaderString(),
which returns a zero-terminated char-string in UTF8-format with these C-style-defs
(in my current C-emitting approach, such C-style header-infos are generated anyways as a "by-product" of the compiling-process).

Most COM-Dlls are able to "describe themselves" (aside from a few excpetions, which have the *.tlb externally) - 
and I like it very much, when we could "keep this up, even on Linux" (using the just described C-style-header-returning Flat-Export).

Olaf

**Comment by bclothier** [comment]
Just for consideration, linking [the SO thread about equivalent of COM on *nix](https://stackoverflow.com/questions/3063321/is-there-an-equivalent-to-com-on-nix-systems-if-not-what-was-the-nix-approa).

Also, as a note of interest the last answer mentions CORBA, which defines a subset of the IDL language used by the MIDL compiler.

**Comment by vbRichClient** [comment]
I did not mean the "remoting-part" of COM (aka "COM+" or formerly as "DCOM") which CORBA tries to mimick.

I meant the COM-ABI (and bindings to a COM-Class at binary level).
- Class-UDT-wise (the VTable-sequence of function-pointers)
- and the "calling" of these function-pointers: HResult = SomeMethod(thisPtr, ...)

And the easiest way to "not alienate" e.g. Python- or LUA-users - 
(in case they want to "bind" to a twinBasic ClassLibrary on Linux), 
is not "a typelib" or "an IDL", but a simple C-Header 
(which potential Authors who plan to write Bindings for twinBasic can immediately understand and use).

Olaf

**Comment by bclothier** [comment]
Olaf, 

I didn't share the this as a suggestion to implement those technologies. In fact, because the equivalent technology's usage seems quite fragmented in *nix, it might not help tB. 

However, if you use the MIDL compiler to build the type library, you'll observe it both generates a header file and also the stub c file which takes care of housekeeping details such as marshaling data types. As I mentioned, COBRA is a subset of the IDL technology that Microsoft uses to generate the type library and/or headers & stubs. MIDL compiler has several Microsoft extensions to the IDL language which makes it easier to build COM stuff regardless of whether remoting is needed or not. 

Will the non-Windows clients benefit from having a stub file in addition to header file? I don't know. Will they similarly benefit from having an equivalent of IDL to help generate the necessary stubs? I also don't know and that also depends on the answer to the first question -- if they must do some housekeeping on the client side then just the header file would not be sufficient as we don't want them getting frustrated when they use an object and it fails because memory wasn't allocated properly or something.

**Comment by wqweto** [comment]
Or the compiler could generate typelib info as data in JSON format so that anyone could generate custom FFI binding in whatever format needed -- .idl, .tlb, .h, .lua or auto-generated C# pinvokes.

**Comment by vbRichClient** [comment]
@bclothier 
> if they must do some housekeeping on the client side then just the header file would not be sufficient

A Header which describes the Class-Structs (and some Enums) will be sufficient, because:
- the instancing of a given Class (its mem-allocation) will happen behind the cFactory-Class (within the *.dll or *.so)
- and the releasing of class-instance-memory, will be ensured by "the class-instance itself"

And don't get me wrong - twinBasic-Dlls as currently produced for Windows -
already inlcude "classic TypeInfo-resources"... why should we take that away?


@wqweto 
Regarding typeinfo in *.h format - that's (IMO) what would be welcomed by most of the languages,
currently popular on the "other side" (Go, Rust, Python, Lua, Freebasic) - all have pre-existing tools to deal with C-Headers nicely in their C-Bindings.

But yes, JSON would be another nice (more generic) description in "plain, human-readable text".
It could be returned by the same function, which should then probably be named:
GetTypeInfoHeaderString(boolAsJSON)  <- boolAsJson left at 0 would then hand out the C-Header.

And in case JSON is returned, this output could then (in addition) include the Help-Text for each Class and Method -
(directly retrieved from the "doc-like comment-attributes" entered in the Editor).
It'd be quite easy to build a documentation-webpage around such a "JSON-output-with-help-text"
(directly from the Lib-Binary, e.g. when it is sitting there on a "NPM-like" acting Repo-Server, waiting for a download).

I love, that COM-Dlls (on Windows) can describe themselves - but always hated that *.tlbs were "binary" - then requiring tools (like OleView) - just to "take a fast peek" (e.g. to look up at which position certain VTable-Members were placed).

As for "Variants, SafeArrays and BSTRs" (and their proper passing into Class-MethodCalls on e.g. Linux) - 
these could be "made easier accessable" by lightweight COM-instances for client-consumers (or wrapper authors) there -
by being retrievable behind methods of the cFactory-Class, every twinProduced Class-Lib exposes...
(using proper constructor-args, to "map-to" or "convert-from" existing "native-allocations").
 
Olaf

**Comment by wqweto** [comment]
Btw, can C# import .h files or is there an automated tool to convert .h files to platform invoke imports?

**Comment by bclothier** [comment]
In C#'s case, the easiest path would be just to use `tlbimp.exe` and import the types from a type library so it wouldn't need to use the C header files. Then it can just use COM types as if they are just another .NET type.

For platform invokes, I am not aware of an automated system to convert the functions and define the p/invoke.

**Comment by vbRichClient** [comment]
>In C#'s case, the easiest path would be just to use tlbimp.exe and import the types from a type library...

Yep, and in case the wrapper.dlls which are produced by TlbImp.exe are not working as they should (for C# on Linux) - 
one could of course generate "normal C#-code" from these MS-IL-based wrapper-binaries as well...
(via ILSpy, Reflector or whatever tools are currently at hand, to "decompile" a .NET-binary).

Olaf

**Comment by bclothier** [comment]
FWIW, here's a different solution, not necessarily better or worse. COM _does_ support parameterized construction but only in a very limited fashion. This is done by using [`IObjectConstruct` interface](https://docs.microsoft.com/en-us/windows/win32/api/comsvcs/nn-comsvcs-iobjectconstruct), which allows a string to be passed in as the construction. 

There's a [sample](https://docs.microsoft.com/en-us/windows/win32/cossdk/using-an-object-constructor-string-to-construct-a-component) of how it can be used to obtain a parameter and thus construct the object from COM. 

This is not as flexible as the true parameterized construction. However, tB can provide means to send in binary data. Example:

Implementing library:
```
Public Type FooConstructionArguments
  Bar As String
  Fizz As Long
  Buzz As Double
End Type

Public Class Foo
  Public Sub New(Parameters As FooConstructionArguments)
    'Do something with the the parameters...
  End Sub
End Class
```

Calling VB/VBA client:
```
Private Arguments As FooConstructionArguments
With Arguments
  ... 'Initialize the values
End With

Foo.Construct(EncodeUDT(Arguments))
```

where the `EncodeUDT` is a tB-defined function that will convert a UDT into a byte array / string and dispatch it. The tB compiler can then expose `Construct` method as if the `Foo` had implemented the `IObjectConstruct`. We unfortunately cannot use `New` since VB/VBA won't call the `IObjectConstruct` that way but at least that would be accessible from COM clients. 
```
